File: Python/54.Spiral Matrix.py
#把矩阵(m*n)中的数按照顺时针顺序输出
class Solution(object):
    def spiralOrder(self, matrix):
        m = len(matrix)
        if m==0:
            return []
        n = len(matrix[0])
        i , j , x ,c ,l = 0 , 0 , 0 , 0 , []
        while x<m*n:
            while x<m*n and j<n-c:
                l.append(matrix[i][j])
                j , x = j+1 , x+1
            j , i =  j-1 , i+1
            while x<m*n and i<m-c:
                l.append(matrix[i][j])
                i , x = i+1 , x+1
            i , j=  i-1 , j-1
            while x<m*n and j>=c:
                l.append(matrix[i][j])
                j , x = j-1 , x+1
            j ,i = j+1, i-1
            while x<m*n and i>c:
                l.append(matrix[i][j])
                i , x = i-1 , x+1
            i ,j = i+1 , j+1
            c += 1
        return l

s = Solution()
print(s.spiralOrder([]))
print(s.spiralOrder([[1],[2]]))
print(s.spiralOrder([[]]))
print(s.spiralOrder([[1,2],[3,4]]))
print(s.spiralOrder([[1,2],[3,4],[5,6],[7,8],[9,10],[11,12],[13,14],[15,16]]))

File: C++/Array/54.Spiral Matrix.cpp
/*
返回二维数组的螺旋形态。
*/

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	vector<int> spiralOrder(vector<vector<int>>& matrix) {
		int m = matrix.size();
        if(m==0)    
        	{vector<int> re;    return re;}
        int n = matrix[0].size(),pos=0,rowstart=0,colstart=0;
		vector<int> result(m--*n--);
		while (rowstart<=m && colstart<=n)
		{
			for (int j = colstart; j <= n; ++j)
			{
				result[pos++] = matrix[rowstart][j];
			}
			++rowstart;
			for (int i = rowstart; i <= m; ++i)
			{
				result[pos++] = matrix[i][n];
			}
			--n;
			if(rowstart<=m)//比如只有1列的情形，如果不判断会多写一列
			{
				for (int j = n; j >= colstart; --j)
				{
					result[pos++] = matrix[m][j];
				}
			
			}
			--m;
			if(colstart<=n)
			{
				for (int i = m;i >= rowstart; --i)
				{
					result[pos++] = matrix[i][colstart];
				}
			}
			++colstart;
		}
		return result;  
    }
};

int main()
{
	Solution s;
	vector<vector<int>> matrix{ { 1, 2, 3, 4 }, { 5, 6, 7, 8 }, { 9, 10, 11, 12 }, {13,14,15,16} };
	//vector<vector<int>> matrix{ { 1, 2, 3 }, { 4, 5, 6 }, { 7, 8, 9 } };
	//vector<vector<int>> matrix{ { 1, 2 }, { 3, 4 }, { 5, 6 }, { 7, 8 }, { 9, 10 }, {11,12} };
	//vector<vector<int>> matrix{ {1,2,3,4,5,6,7,8} };
	vector<int> out;
	out = s.spiralOrder(matrix);
	for(int i=0; i < out.size(); ++i)
	{
		cout<<out[i]<<endl;
	}
}

